tools/pythonpkg/tests/fast/test_replacement_scan.py

```diff
@@ -125,6 +146,125 @@ def test_replacement_scan_fail(self):
         ):
             con.execute("select count(*) from random_object").fetchone()
 
+    @pytest.mark.parametrize(
+        'df_create',
+        [
+            from_pandas,
+            from_arrow,
+        ],
+    )
+    def test_cte(self, duckdb_cursor, df_create):
+        df = df_create()
+        rel = duckdb_cursor.sql("with cte as (select * from df) select * from cte")
+        res = rel.fetchall()
+        assert res == [(1,), (2,), (3,)]
+
+        df = df_create()
+        query = """
+            WITH cte as (select * from df)
+            select * from (
+                WITH cte as (select * from df)
+                select * from (
+                    WITH cte as (select * from df)
+                    select (
+                        select * from cte
+                    ) from cte
+                )
+            )
+        """
+        rel = duckdb_cursor.sql(query)
+        duckdb_cursor.execute("create table df as select * from range(4, 7)")
+        res = rel.fetchall()
+        """
+        select (
+            select * from cte
+        ) from cte
+        This will select the first row from the cte, and do this 3 times since we added 'from cte', and cte has 3 tuples
+        """
+
+        if df_create == from_arrow:
+            # Because the RecordBatchReader is destructive, it's empty after the first scan
+            # But we reference it multiple times, so the subsequent reads have no data to read
+            assert res == [(None,), (None,), (None,)]
+        else:
+            assert res == [(1,), (1,), (1,)]
+
+    def test_cte_with_joins(self, duckdb_cursor):
+        df = pd.DataFrame({'a': [1, 2, 3]})
+        query = """
+            WITH cte1 AS (
+                SELECT * FROM df
+            ),
+            cte2 AS (
+                SELECT * FROM df
+                WHERE a > 1
+            ),
+            cte3 AS (
+                SELECT * FROM df
+                WHERE a < 3
+            )
+            SELECT * FROM (
+                SELECT 
+                    cte1.*, 
+                    cte2.a AS cte2_a,
+                    subquery.a AS cte3_a
+                FROM cte1
+                JOIN cte2 ON cte1.a = cte2.a
+                JOIN (
+                    SELECT 
+                        df.*, 
+                        cte3.a AS cte3_a
+                    FROM df
+                    JOIN cte3 ON df.a = cte3.a
+                ) AS subquery ON cte1.a = subquery.a
+            ) AS main_query
+            WHERE main_query.a = 2
+        """
+        rel = duckdb_cursor.sql(query)
+        duckdb_cursor.execute("create table df as select * from range(4, 7)")
+        res = rel.fetchall()
+        assert res == [(2, 2, 2)]
+
+    def test_cte_at_different_levels(self, duckdb_cursor):
+        df = pd.DataFrame({'a': [1, 2, 3]})
+        query = """
```

Comment:
* Can you add a test with scalar subqueries?
* Can you add a test with a recursive CTE?
* Can you add a test where the CTE name is the same as the dataframe name, e.g. `WITH df AS (SELECT a+1 FROM df) SELECT * FROM df;`

Reply:
Thanks for the suggestions!